R/teal_lockfile.R

```diff
@@ -0,0 +1,114 @@
+#' Generate lockfile for application reproducibility
+#'
+#' This function is invoked during [teal::init] to create `renv`-compatible lockfile for use within the application.
+#'
+#' The function leverages [renv::snapshot()], which offers multiple methods for lockfile creation.
+#'
+#' - User-specified:
+#'     - **Pre-computed lockfile**: Users can provide their own pre-computed lockfile by specifying the path via
+#'     `teal.renv.lockfile` option. Automatic lockfile computation is skipped in such case.
+#' - Automatically computed:
+#'     - **Working directory lockfile**: If `teal.renv.lockfile` is not set, `teal` will, by default, create an
+#'      `implicit` type lockfile that uses `renv::dependencies()` to detect all R packages in the current project's
+#'      working directory.
+#'     - **`DESCRIPTION`-based lockfile**: To generate a lockfile based on a `DESCRIPTION` file in your working
+#'     directory, set `renv::settings$snapshot.type("explicit")`. The naming convention for `type` follows
+#'     `renv::snapshot()`. For the `"explicit"` type, refer to `renv::settings$package.dependency.fields()` for the
+#'     `DESCRIPTION` fields included in the lockfile.
+#'     - **Custom files-based lockfile**: To specify custom files as the basis for the lockfile, set
+#'     `renv::settings$snapshot.type("custom")` and configure the `renv.snapshot.filter` option.
+#'
+#' @section lockfile usage:
+#' After creating the lockfile, you can restore the application environment using `renv::restore()`.
+#'
+#' @seealso [renv::snapshot()], [renv::restore()].
+#'
+#' @return Nothing. This function is executed for its side effect of creating a lockfile used in the `teal` application.
+#'
+#' @keywords internal
+teal_lockfile <- function() {
+  lockfile_path <- "teal_app.lock"
+  # If user has setup the file, there is no need to compute a new one.
+  user_lockfile <- getOption("teal.renv.lockfile", "")
+  if (!identical(user_lockfile, "")) {
+    if (file.exists(user_lockfile)) {
+      file.copy(user_lockfile, lockfile_path)
+      return(invisible(NULL))
+    } else {
+      stop("lockfile provided through options('teal.renv.lockfile') does not exist.")
+    }
+  }
+
+  if (!(is_in_test() || is_r_cmd_check())) {
+    old_plan <- future::plan()
+    # If there is already a parallel (non-sequential) backend, reuse it.
+    if (inherits(old_plan, "sequential")) {
+      future::plan(future::multisession, workers = 2)
+    }
+
+    lockfile_task <- ExtendedTask$new(create_renv_lockfile)
+    lockfile_task$invoke(close = inherits(old_plan, "sequential"), lockfile_path)
```

Comment:
Why not to restore an old plan after this line?

Reply:
I see now, you followed "Avoid changing the future strategy" section [from here](https://cran.r-project.org/web/packages/future/vignettes/future-7-for-package-developers.html)

Problem is that in the vignette they undo plan to the previous regardless of the future evaluation time. In your case, plan is changed after `renv::snapshot` is run which IMO is wrong. It is wrong because:
- `create_renv_lockfile()` can take longer time which spans possibly other `future` invokations which were planned differently
- Restoring `plan(old_plan)` in `teal_lockfile()` `on.exit` seems to be better

Reply:
I was worried not to change the plan before the task is finished. 
From the vignette we can only _assume_ that even though you run `on.exit(old_plan)` the task is not terminated if the plan is changed. 

We can test it out. If the functionality keeps running the renv task, even though we have `on.exit(old_plan)` included, that would be a great simplification of the code, and would probably mean we can get rid of `promises` package that I mainly used for closing the plan on a follow-up after the task is finished

```r
    promises::then(promise, onFulfilled = function() {
      future::plan(future::sequential)
    })
```

Reply:
> I was worried not to change the plan before the task is finished.

Here we are not sure either if any tasks are finished or not.

```
if (inherits(old_plan, "sequential")) {
  future::plan(future::multisession, workers = 2)
}
```

Reply:
Okey @m7pr , you are right! You can't change plan while worker is being executed.

```r
library(future)
plan(multisession)

a %<-% {
  Sys.sleep(10)
  1L
}

plan(sequential)
b %<-% {
  Sys.sleep(2)
  2L
}

c %<-% {
  3L
}

a
```

```

Error: MultisessionFuture (<none>) failed to receiving message from cluster RichSOCKnode #1 (PID 39539 on localhost ‘localhost’). The reason reported was ‘Connection to the worker is corrupt’. Post-mortem diagnostic: A process with this PID exists, which suggests that the localhost worker is still alive. The socket connection to the worker of MultisessionFuture future (<none>) is lost or corrupted: Connection (connection: index=4, description="NA", class="NA", mode="NA", text="NA", opened="NA", can read="NA", can write="NA", id=326, raw_id="<pointer: 0x146>") is no longer valid. There is currently no registered R connection with that index 4. As an example, this may happen if base::closeAllConnections() have been called, for instance via base::sys.save.image() which in turn is called if the R session (pid 39502) is forced to terminate
```